File: 0121/one_pass.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    int ret = 0;
    if (prices.size() == 0) return ret;
    int min_price = prices[0];
    for (size_t i = 1; i < prices.size(); ++i) {
      min_price = min(min_price, prices[i]);
      ret = max(ret, prices[i] - min_price);
    }
    return ret;
  }
};

int main() {
  Solution sol;
  vector<int> prices;

  prices = {7,1,5,3,6,4};
  assert(sol.maxProfit(prices) == 5);

  prices = {7,6,4,3,1};
  assert(sol.maxProfit(prices) == 0);

  prices = {};
  assert(sol.maxProfit(prices) == 0);
}

File: 0121/one_pass.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ret = 0
        if len(prices) == 0:
            return ret
        valley, peak = prices[0], prices[0]
        for idx in range(1, len(prices)):
            price = prices[idx]
            if price < peak:
                ret = max(ret, peak - valley)
            peak = price
            valley = min(valley, price)
        ret = max(ret, peak - valley)
        return ret


if __name__ == '__main__':
    sol = Solution()

    prices = [7, 1, 5, 3, 6, 4]
    expected = 5
    assert sol.maxProfit(prices) == expected

    prices = [7, 6, 4, 3, 1]
    expected = 0
    assert sol.maxProfit(prices) == expected

    prices = []
    expected = 0
    assert sol.maxProfit(prices) == expected

    prices = [1, 2]
    expected = 1
    assert sol.maxProfit(prices) == expected
